nni/compression/pytorch/speedup/jit_translate.py

```diff
@@ -461,179 +162,311 @@ def forward(self, obj):
-            , the following parameters is useless, because we already
-            get the size_list and the scale_list by parsing the cpp_nodes.
-            """
-            return torch.nn.functional.upsample_nearest(args[0],
-                                                        size=self.size_list, scale_factor=self.scale_list)
-    c_node = node.key_node
-    inputs = list(c_node.inputs())
-    size_list_node = inputs[1].node()
-    scale_list_node = inputs[2].node()
-    size_list = None
-    scale_list = None
+    """
+
+    func: Callable
+    positional: List[Any]
+    keyword: Dict[str, Any]
+    undetermined: List[Union[int, str]]
+    special_treat: Dict[Union[int, str], Callable]
+
+    def __new__(cls, func: Callable, positional: List[Any], keyword: Dict[str, Any],
+                undetermined: List[Union[int, str]], special_treat: Dict[Union[int, str], Callable]):
+        if not callable(func):
+            raise TypeError("the 'func' argument must be callable")
+
+        self = super(func_adapter, cls).__new__(cls)
+
+        self.func = func
+        self.positional = positional
+        self.keyword = keyword
+        self.undetermined = undetermined
+        self.special_treat = special_treat
+        return self
+
+    def __call__(self, /, *args):
+        assert len(args) >= len(self.undetermined)
+        if len(args) > len(self.undetermined):
+            logger.warning('throw some args away when calling the function \'%s\'', self.func.__name__)
+        for i, p in enumerate(self.undetermined):
+            v = args[i]
+            if type(p) is int:
+                self.positional[p] = v
+            else:
+                self.keyword[p] = v
+
+        for p, fs in self.special_treat.items():
+            if isinstance(p, int):
+                for f in fs: self.positional[p] = f(self.positional[p])
+            else:
+                for f in fs: self.keyword[p] = f(self.keyword[p])
+        result = self.func(*self.positional, **self.keyword)
+        if type(result) is int: # turn result of 'size' into tensor
+            result = torch.as_tensor([result], dtype=torch.long)
+        return result
+
+# There are some types that will be convert into enums after jit.
+# So we should recover them back:
+#   device, dtype, layout, memory_format, qscheme, qengine, dispatchkey
+
+enum_to_dtype_names = {
+    0: 'uint8',
+    1: 'int8',
+    2: 'int16',
+    3: 'int32',
+    4: 'int64',
+    5: 'float16',
+    6: 'float32',
+    7: 'float64',
+    8: 'complex32',
+    9: 'complex64',
+    10: 'complex128',
+    11: 'bool',
+    12: 'qint8',
+    13: 'quint8',
+    14: 'qint32',
+    15: 'bfloat16',
+    16: 'quint4x2',
+    17: 'quint2x4',
+}
+enum_to_dtype_dict = {}
+for enum_value, dtype_name in enum_to_dtype_names.items():
+    if hasattr(torch, dtype_name):
+        enum_to_dtype_dict[enum_value] = getattr(torch, dtype_name)
+def dtype_trans(ivalue: Union[int, torch.dtype]):
```

Comment:
one or two blank lines before function maybe better.
just a personal preference, using `int | torch.dtype` instead of `Union[int, torch.dtype]`

Reply:
commented by zhe liu, the python version in our test env is old and do not support this grammar(3.10+)

Reply:
https://peps.python.org/pep-0604/

Reply:
yes, but I see that you add this line `from __future__ import annotations`

Reply:
> yes, but I see that you add this line `from __future__ import annotations`

`__future__` makes it work at runtime.
But it may fail the static type checking, depending on the checker's implementation.